contracts/pcv/PCVDepositAggregator.sol

```diff
@@ -0,0 +1,444 @@
+    /// 1. check if the contract has enough in the buffer to cover the withdrawal. if so, just use this
+    /// 2. if not, calculate what the ideal underlying amount should be for each pcv deposit *after* the withdraw
+    /// 3. then, cycle through them and withdraw until each has their ideal amount (for the ones that have overages)
+    /// Note this function will withdraw all of the overages from each pcv deposit, even if we don't need that much to
+    /// actually cover the transfer! This is intentional because it costs the same to withdraw exactly how much we need
+    /// vs the overage amount; the entire overage amount should be moved if it is the same cost as just as much as we need.
+    function withdraw(address to, uint256 amount) external override onlyPCVController whenNotPaused {
+        uint256 aggregatorBalance = balance();
+
+        if (aggregatorBalance > amount) {
+            IERC20(token).safeTransfer(to, amount);
+            return;
+        }
+        
+        uint256[] memory underlyingBalances = _getUnderlyingBalances();
+        uint256 totalUnderlyingBalance = underlyingBalances.sum();
+        uint256 totalBalance = totalUnderlyingBalance + aggregatorBalance;
+
+        require(totalBalance >= amount, "Not enough balance to withdraw");
+
+        // We're going to have to pull from underlying deposits
+        // To avoid the need from a rebalance, we should withdraw proportionally from each deposit
+        // such that at the end of this loop, each deposit has moved towards a correct weighting
+        uint256 amountNeededFromUnderlying = amount - aggregatorBalance;
+        uint256 totalUnderlyingBalanceAfterWithdraw = totalUnderlyingBalance - amountNeededFromUnderlying;
+
+        // Next, calculate exactly the desired underlying balance after withdraw
+        uint[] memory idealUnderlyingBalancesPostWithdraw = new uint[](pcvDepositAddresses.length());
+        for (uint256 i=0; i < pcvDepositAddresses.length(); i++) {
+            idealUnderlyingBalancesPostWithdraw[i] = totalUnderlyingBalanceAfterWithdraw * pcvDepositWeights[pcvDepositAddresses.at(i)] / totalWeight;
+        }
+
+        // This basically does half of a rebalance.
+        // (pulls from the deposits that have > than their post-withdraw-ideal-underlying-balance)
+        for (uint256 i=0; i < pcvDepositAddresses.length(); i++) {
+            address pcvDepositAddress = pcvDepositAddresses.at(i);
+            uint256 actualPcvDepositBalance = underlyingBalances[i];
+            uint256 idealPcvDepositBalance = idealUnderlyingBalancesPostWithdraw[i];
+
+            if (actualPcvDepositBalance > idealPcvDepositBalance) {
+                // Has post-withdraw-overage; let's take it
+                uint256 amountToWithdraw = actualPcvDepositBalance - idealPcvDepositBalance;
+                IPCVDeposit(pcvDepositAddress).withdraw(address(this), amountToWithdraw);
```

Comment:
This can withdraw too much, right?

Reply:
amountToWithdraw is unrelated to the originally requested `amount`

Reply:
Nope. See the note at the top of this function:

```
    /// Note this function will withdraw all of the overages from each pcv deposit, even if we don't need that much to
    /// actually cover the transfer! This is intentional because it costs the same to withdraw exactly how much we need
    /// vs the overage amount; the entire overage amount should be moved if it is the same cost as just as much as we need.
    ```

Reply:
If you pair this with deposit, we have the same security concerns as before as atomic withdraw + deposit is equivalent to rebalance

Reply:
Aaaaaah you're right. I'll remove it.

Reply:
withdraw is onlyPCVController though